lib/python/rose/popen.py

```diff
@@ -271,6 +272,57 @@ def which(name):
             if os.access(file_name, os.F_OK | os.X_OK):
                 return file_name
 
+    async def run_bg_async(self, *args, **kwargs):
+        """Provide a Rose-friendly interface to subprocess.Popen.
+
+        Return a subprocess.Popen object.
+
+        If kwargs["stdin"] is a str, turn it into subprocess.PIPE.
+        However, it cannot communicate stdin to the Popen object,
+        because the Popen.communicate() method implies a wait().
+
+        """
+        for key in ["stdout", "stderr"]:
+            if kwargs.get(key) is None:
+                kwargs[key] = asyncio.subprocess.PIPE
+        stdin = kwargs.get("stdin")
+        if isinstance(stdin, str):
+            kwargs["stdin"] = asyncio.subprocess.PIPE
+        elif stdin is None:
+            kwargs["stdin"] = open(os.devnull)
+        self.handle_event(RosePopenEvent(args, stdin))
+        sys.stdout.flush()
+        try:
+            if kwargs.get("shell"):
+                proc = Popen(args[0], **kwargs)
+            else:
+                command = ' '.join(args)
```

Comment:
`asyncio.create_subprocess_shell` appears to be completely independent of the `subprocess` library, it is roughly equivalent to using `Popen` with the `shell=True` argument which has safety concerns which we will have to address.

The best I can come up with is this, it's safe though may lead to some cryptic errors when misused:

```python
command = ' '.join(map(shlex.quote, args))
```

```python
>>> args = ['sleep', '1', ';', 'echo', 'foo']
>>> ' '.join(map(shlex.quote, args))
"sleep 1 ';' echo foo"

>>> args = ['sleep', '1;', 'echo foo']
>>> ' '.join(map(shlex.quote, args))
"sleep '1;' 'echo foo'"
```

Reply:
:worried: